front/src/app/core/login/services/auth.service.ts

```diff
@@ -0,0 +1,407 @@
+import { Injectable, inject } from '@angular/core';
+import { Router } from '@angular/router';
+import { HttpClient } from '@angular/common/http';
+import {BehaviorSubject, firstValueFrom, Observable, of} from 'rxjs';
+import { MatDialog } from '@angular/material/dialog';
+import {
+  Auth,
+  User as FirebaseUser,
+  onAuthStateChanged,
+  signInWithPopup,
+  GoogleAuthProvider,
+  GithubAuthProvider,
+  signInWithEmailLink,
+  isSignInWithEmailLink,
+  sendSignInLinkToEmail,
+  fetchSignInMethodsForEmail,
+  setPersistence,
+  browserLocalPersistence,
+  signOut, authState
+} from '@angular/fire/auth';
+import {UserStateService} from '../../services/user-services/user-state.service';
+import {User} from '../../models/user.model';
+import {environment} from '../../../../environments/environment.development';
+import {AuthErrorDialogComponent} from '../../../shared/auth-error-dialog/auth-error-dialog.component';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AuthService {
+  private currentUserSubject = new BehaviorSubject<User | null>(null);
+
+  private auth = inject(Auth);
+  private http = inject(HttpClient);
+  private router = inject(Router);
+  private dialog = inject(MatDialog);
+  private userState = inject(UserStateService);
+
+  user$ = new BehaviorSubject<FirebaseUser | null>(null);
```

Comment:
todo: bad practice de mettre un BehaviorSubject en public comme ça

N'importe où dans l'app tu peux faire un authService.user$.next(...) et pourrir ton état. Du coup c'est compliqué de comprendre le flow de l'application.

le mieux c'est de mettre le Subject private et d'avoir un getter qui expose un Observable

```suggestion
  private userSubject$ = new BehaviorSubject<FirebaseUser | null>(null);
  
  get user$() {
      return this.userSubject$.asObservable();
  }
```

Et si t'as besoin de mettre à jour depuis l'extérieur du service un jour, tu ajoutes une méthode setFirebaseUser() ou un truc du genre

Exemple qui détail plus le pattern : https://angular.fr/services/behavior-subject